src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs

```diff
@@ -15,37 +17,114 @@ class TimeoutCancellationTokenSourceWrapper : IDisposable
     {
         public TimeSpan Timeout { get; set; }
         CancellationTokenSource _timeoutCts = new CancellationTokenSource();
+        CancellationToken _linkedToken;
 
-        public TimeoutCancellationTokenSourceWrapper() => Timeout = System.Threading.Timeout.InfiniteTimeSpan;
+#if DEBUG
+        bool _isRunning;
+#endif
+
+        public TimeoutCancellationTokenSourceWrapper() => Timeout = InfiniteTimeSpan;
 
         public TimeoutCancellationTokenSourceWrapper(TimeSpan timeout) => Timeout = timeout;
 
         /// <summary>
         /// Set the timeout on the wrapped <see cref="CancellationTokenSource"/>
         /// and make sure that it hasn't been cancelled yet
         /// </summary>
-        public CancellationToken Start()
+        /// <param name="cancellationToken">An optional cancellation token that will be linked with the
+        /// contained <see cref="CancellationTokenSource"/></param>
+        /// <returns>The <see cref="CancellationToken"/> from the wrapped <see cref="CancellationTokenSource"/></returns>
+        public CancellationToken Start(CancellationToken cancellationToken = default)
         {
-            _timeoutCts.CancelAfter(Timeout);
-            if (_timeoutCts.IsCancellationRequested)
+#if DEBUG
+            Debug.Assert(!_isRunning);
+#endif
+
+            // If we already wrap a linked token source or we want to wrap a linked token source and the token to
+            // link isn't the same one that's already linked we always have to create a new CancellationTokenSource
+            if ((_linkedToken.CanBeCanceled || cancellationToken.CanBeCanceled) && _linkedToken != cancellationToken)
             {
                 _timeoutCts.Dispose();
-                _timeoutCts = new CancellationTokenSource(Timeout);
+                _timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
```

Comment:
nit: can rename _timeoutCts to just _cts as we're within the context of the wrapper